Add `bySemanticsIdentifier` finder for finding by identifier

admosity opened the pull request:
## Add `bySemanticsIdentifier` finder for finding by identifier

### Description

This pull request introduces a new finder, `CommonFinders.bySemanticsIdentifier`, to the Flutter testing framework. This finder allows developers to locate `Semantics` widgets based on their `identifier` property, enhancing the precision and flexibility of widget tests.

### Motivation

Establish a consistent and reliable method for locating elements in integration and end-to-end (e2e) tests. Unlike `label` or `key`, which may carry functional significance within the application, the `identifier` is purely declarative and does not impact functionality. Utilizing the `identifier` for finding semantics widgets ensures that tests can target specific elements without interfering with the app's behavior, thereby enhancing test reliability, maintainability, and reusability across testing frameworks.

### Changes

- **semantics.dart**
  - Updated documentation to mention that `identifier` can be matched using `CommonFinders.bySemanticsIdentifier`.
  
- **finders.dart**
  - Added the `bySemanticsIdentifier` method to `CommonFinders`.
  - Supports both exact string matches and regular expression patterns.
  - Includes error handling to ensure semantics are enabled during tests.
  
- **finders_test.dart**
  - Added tests to verify that `bySemanticsIdentifier` correctly finds widgets by exact identifier and regex patterns.
  - Ensures that the finder behaves as expected when semantics are not enabled.

### Usage

Developers can use the new finder in their tests as follows:

```dart
// Exact match
expect(find.bySemanticsIdentifier('Back'), findsOneWidget);

// Regular expression match
expect(find.bySemanticsIdentifier(RegExp(r'^item-')), findsNWidgets(2));
```

admosity committed:
feat(finders): Add `bySemanticsIdentifier` finder for locating Semantics widgets

- Introduce `CommonFinders.bySemanticsIdentifier` to enable finding Semantics widgets by their `identifier`.
- Update `semantics.dart` documentation to include usage of the new finder.
- Add comprehensive tests in `finders_test.dart` to verify functionality with both exact strings and regular expressions.

Sample usage:
```dart
expect(find.bySemanticsIdentifier('Back'), findsOneWidget);
expect(find.bySemanticsIdentifier(RegExp(r'^item-')), findsNWidgets(2));
```

navaronbracke commented on packages/flutter_test/test/finders_test.dart:
```diff
@@ -299,6 +299,45 @@ void main() {
       semanticsHandle.dispose();
     });
+
+    testWidgets('finds Semantically labeled widgets by identifier',
```
Can you also test the StateError when semantics is disabled?

navaronbracke commented on packages/flutter_test/lib/src/finders.dart:
```diff
@@ -579,8 +579,52 @@ class CommonFinders {
+    }
 }
 
+
```
Redundant newline

admosity committed:
feat(finders): add semantics enabled test and remove redundant line

navaronbracke commented on packages/flutter_test/lib/src/finders.dart:
```diff
@@ -579,6 +579,49 @@ class CommonFinders {
     );
   }
+
+
```
There is an additional newline here that can be removed.

navaronbracke commented on packages/flutter_test/test/finders_test.dart:
```diff
@@ -299,6 +299,50 @@ void main() {
+      expect(find.bySemanticsIdentifier(RegExp(r'^item-')), findsNWidgets(2));
+      semanticsHandle.dispose();
+    });
+
```
There is an additional newline here that can be removed.

navaronbracke commented on packages/flutter_test/test/finders_test.dart:
```diff
@@ -299,6 +299,50 @@ void main() {
+
+    testWidgets('Throws StateError if semantics are not enabled (bySemanticsIdentifier)',
+        (WidgetTester tester) async {
+      expect(() => find.bySemanticsIdentifier('Add'), throwsStateError);
```
Can you also test the StateError's message, using the `having()` matcher?

admosity replied:
Something like this?

```dart
    testWidgets('Throws StateError if semantics are not enabled (bySemanticsIdentifier)',
        (WidgetTester tester) async {
      expect(
        () => find.bySemanticsIdentifier('Add'),
        throwsA(isA<StateError>().having((e) => e.message, 'message', contains('Semantics are not enabled')))
      );
    }, semanticsEnabled: false);
```

admosity replied:
Copied the test from above, shall I update that as well while I am in here?

```dart
    testWidgets('Throws StateError if semantics are not enabled',
        (WidgetTester tester) async {
      expect(() => find.bySemanticsLabel('Add'), throwsStateError);
    }, semanticsEnabled: false);
```

admosity replied:
> Something like this?
> 
> ```dart
>     testWidgets('Throws StateError if semantics are not enabled (bySemanticsIdentifier)',
>         (WidgetTester tester) async {
>       expect(
>         () => find.bySemanticsIdentifier('Add'),
>         throwsA(isA<StateError>().having((e) => e.message, 'message', contains('Semantics are not enabled')))
>       );
>     }, semanticsEnabled: false);
> ```

Gone ahead and applied this for now...

navaronbracke replied:
Yes that is correct

admosity committed:
fix(finders): test the StateError message contents and lint

nate-thegrate commented on packages/flutter_test/lib/src/finders.dart:
```diff
@@ -579,6 +579,48 @@ class CommonFinders {
+  ///
+  /// If the `skipOffstage` argument is true (the default), then this skips
+  /// nodes that are [Offstage] or that are from inactive [Route]s.
+  Finder bySemanticsIdentifier(Pattern identifier, {bool skipOffstage = true}) {
+      if (!SemanticsBinding.instance.semanticsEnabled) {
+        throw StateError('Semantics are not enabled. '
```
It looks like the entire function body has 1 extra indent.

nate-thegrate reviewed (commented):
Thanks for adding this, looks fantastic overall!

Just have a bunch of whitespace suggestions :)

admosity committed:
style: apply style guide suggestions

Co-authored-by: Nate Wilson <nathan.wilson1232@gmail.com>

admosity committed:
style: indent

nate-thegrate reviewed (approved):
LGTM, thanks for making this addition!

chunhtai commented on packages/flutter_test/lib/src/finders.dart:
```diff
@@ -579,6 +579,49 @@ class CommonFinders {
+  /// If the `skipOffstage` argument is true (the default), then this skips
+  /// nodes that are [Offstage] or that are from inactive [Route]s.
+  Finder bySemanticsIdentifier(Pattern identifier, {bool skipOffstage = true}) {
+    if (!SemanticsBinding.instance.semanticsEnabled) {
```
Can you refactor some code with the bySemanticsLabel, a big portion is duplicated

admosity replied:
Done 👍🏻

admosity committed:
refactor: apply DRY to semantics label and identifier finders

- Refactored `bySemanticsLabel` and `bySemanticsIdentifier` methods to use a common private method `_bySemanticsProperty`.
- Improved code readability and reduced duplication by extracting common logic for semantics property retrieval.
- Updated parameter names for consistency and clarity.

admosity committed:
refactor: rename parameters to match docs

Renamed the parameter `pattern` to `label` in the `bySemanticsLabel` method and `pattern` to `identifier` in the `bySemanticsIdentifier` method.

kevmoo commented:
We good here @chunhtai ?

chunhtai reviewed (approved):
LGTM
